src/content_stream.cpp

```diff
@@ -69,7 +79,15 @@ std::string ContentReader::read() {
     return "";
   }
   // Expect \r\n\r\n
-  if (!match("\r\n\r\n")) {
+  auto terminator_len = strlen(HEADER_TERMINATOR);
+  auto before_matched_idx = matched_idx;
+  if (!match(reinterpret_cast<const uint8_t*>(HEADER_TERMINATOR),
```

Comment:
`match()` calls `buffer()` so I'm not sure why you need this chunk of changes.
Can't this just be:

```
if (!match("\r\n\r\n")) {
    return badHeader();
}
```

---

I investigated this further and I think `StringBuffer` is being incorrectly used by tests.

The contract for `Reader::read()` says:

https://github.com/google/cppdap/blob/60fee6e5059dc8a85c32356193192a79f5082ea4/include/dap/io.h#L39-L42

`StringBuffer` returns `0` when the string has been fully read:

https://github.com/google/cppdap/blob/60fee6e5059dc8a85c32356193192a79f5082ea4/src/string_buffer.h#L65

However pipes will block if they're still open and there's more data that may eventually come.

So `match()` should block until the whole pattern is read, or the pipe is closed before matching.

`StringBuffer` could be updated to hold a list of chunk lengths that are returned by `read()`. That way you can still test for partial reads.

I think #50 should be reverted.

Reply:
Ah, great, thanks for clarifying that. I was working around match returning 0 in two different cases.

You mentioned reverting #50. Should that be done as part of this PR or separately?

Reply:
> Ah, great, thanks for clarifying that. I was working around `match` returning 0 in two different cases.
> 
> You mentioned reverting #50. Should that be done as part of this PR or separately?

Doing it as part of the PR (or as a separate base PR) would probably simplify things as it removes the icky `matched_idx` logic.